src/util.rs

```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
+            )
+                [..];
+
+            let rty = Ident::from({
+                let name = if has_brackets {
+                    info.name.replace("[%s]", "")
+                } else {
+                    info.name.replace("%s", "")
+                };
+                name.to_sanitized_upper_case().into_owned()
+            });
+
+            let reg_name = Ident::from({
+                let name = if has_brackets {
+                    info.name.replace("[%s]", "")
+                } else {
+                    info.name.replace("%s", "")
+                };
+                name.to_sanitized_snake_case().into_owned()
+            });
+
+            let length = Ident::from(array_info.dim.to_string());
+
+            quote! {
+                #[doc = #comment]
+                pub #reg_name : [#rty; #length],
+            }
+        },
+    }
 }
 
+
 /// Takes a list of "registers", some of which may actually be register arrays,
-/// and turns it into a new *sorted* (by address offset) list of registers where
-/// the register arrays have been expanded.
-pub fn expand(registers: &[Register]) -> Vec<ExpandedRegister> {
+/// and turn in into alist of registers where the register arrays have been expanded.
+pub fn expand(register: &Register) -> Vec<Register> {
     let mut out = vec![];
 
-    for r in registers {
-        match *r {
-            Register::Single(ref info) => {
-                out.push(
-                    ExpandedRegister {
-                        info: info,
-                        name: info.name.to_sanitized_snake_case().into_owned(),
-                        offset: info.address_offset,
-                        ty: Either::Left(
-                            info.name
-                                .to_sanitized_upper_case()
-                                .into_owned(),
-                        ),
+    match *register {
+        Register::Single(ref _info) => out.push( register.clone() ),
+        Register::Array(ref info, ref array_info) => {
+            let has_brackets = info.name.contains("[%s]");
+
+            let indices = array_info
+                .dim_index
+                .as_ref()
+                .map(|v| Cow::from(&**v))
+                .unwrap_or_else(
+                    || {
+                        Cow::from(
+                            (0..array_info.dim)
+                                .map(|i| i.to_string())
+                                .collect::<Vec<_>>(),
+                        )
                     },
-                )
-            }
-            Register::Array(ref info, ref array_info) => {
-                let has_brackets = info.name.contains("[%s]");
+                );
 
-                let ty = if has_brackets {
-                    info.name.replace("[%s]", "")
+            for (idx, i) in indices.iter().zip(0..) {
+                let name = if has_brackets {
+                    info.name.replace("[%s]", format!("%s({})", idx).as_str())
```

Comment:
why do we want to keep the "%s" in this stage? Is it going to be used afterwards? Why not just directly replace it with the formatted `idx` at this point?